python_modules/libraries/dagster-polars/dagster_polars/io_managers/type_routers.py

```diff
@@ -0,0 +1,173 @@
+        self.context = context
+        self.typing_type = typing_type
+
+    @abstractmethod
+    def match(self) -> bool:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def is_root_type(self) -> bool:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def parent_type(self) -> Any:
+        raise NotImplementedError
+
+    @property
+    def parent_type_router(self) -> "TypeRouter":
+        return resolve_type_router(self.context, self.parent_type)
+
+    @property
+    def is_eager(self) -> bool:
+        if self.is_root_type:
+            return self.typing_type in [Any, type(None), None] or issubclass(
+                self.typing_type, pl.DataFrame
+            )
+        else:
+            return self.parent_type_router.is_eager
+
+    @property
+    def is_lazy(self) -> bool:
+        if self.is_root_type:
+            return issubclass(self.typing_type, pl.LazyFrame)
+        else:
+            return self.parent_type_router.is_lazy
+
+    def dump(self, obj: T, path: "UPath", dump_fn: F_D) -> None:
+        if self.is_root_type:
+            dump_fn(self.context, obj, path)
+        else:
+            self.parent_type_router.dump(obj, path, dump_fn)
+
+    def load(self, path: "UPath", load_fn: F_L) -> T:
+        if self.is_root_type:
+            return load_fn(path, self.context)
+        else:
+            return self.parent_type_router.load(path, load_fn)
+
+
+class TypeRouter(BaseTypeRouter, Generic[T]):
+    """Handles most common standard types."""
+
+    def match(self) -> bool:
+        return self.typing_type in [
+            pl.DataFrame,
+            pl.LazyFrame,
+            Any,
+            type(None),
+            None,
+        ]
+
+    @property
+    def is_root_type(self) -> bool:
+        return True
+
+
+class OptionalTypeRouter(BaseTypeRouter, Generic[T]):
+    """Handles Optional type annotations with a noop if the object is None or missing in storage."""
+
+    def match(self) -> bool:
+        return get_origin(self.typing_type) == Union and type(None) in get_args(self.typing_type)
+
+    @property
+    def is_root_type(self) -> bool:
+        return False
+
+    @property
+    def parent_type(self) -> Any:
+        return get_args(self.typing_type)[0]
+
+    def dump(self, obj: T, path: "UPath", dump_fn: F_D) -> None:
+        if obj is None:
+            self.context.log.warning(f"Skipping saving optional output at {path} as it is None")
+            return
+        else:
+            self.parent_type_router.dump(obj, path, dump_fn)
+
+    def load(self, path: "UPath", load_fn: F_L) -> T:
+        if not path.exists():
+            self.context.log.warning(f"Skipping loading optional input at {path} as it is missing")
+            return None
+        else:
+            return self.parent_type_router.load(path, load_fn)
+
+
+class DictTypeRouter(BaseTypeRouter, Generic[T]):
```

Comment:
I see you have this under test but I'm somewhat confused as to how this actually works without the load function being overwritten. Where is the logic that handles calling `load` on each input partition and assembling that into a dictionary?

Reply:
This happens in the base `UPathIOManager`